### ISSUE TITLE
 Subprojects update needs a --force option

### ISSUE THREAD

**Issue by alatiera** [mesonbuild/meson#7526]
Scenario is as follows, you depend on a subproject stable branch like `glib 2.60`. You bump the minimum dependency to 2.64 and the .wrap file but that isn't reflected in when reconfiguring. Instead it complains cause the there already exists a cached subproject which does not meet the new requirements.

You can't pull or rebase, cause the branches have divergent histories. The only solution is to `rm -rf` the project and reclone or manually checkout. This might work for one subproject and developer checkouts but its impossible to do in CI environments easily and scalable. In gst-build CI we are caching the non-gst subprojects to avoid redownloading the git repos each time which they are *huge* due to ffmpeg, gtk, glib etc. 

The desired behavior would be a way for meson to make sure the current checked out subproject is matching the revision in the .wrap file and is up to date. Something like `meson subprojects update --force` where it would fetch the .wrap file and then `checkout -f` in some way.

**Comment by xclaesse** [comment]
+1 I like that idea. I also would like to have gst-build's manifest feature at some point. Basically I want gst-build's git-update.py to be completely replaced by meson subprojects update, but that's a long term goal.

**Comment by tp-m** [comment]
Maybe we don't need a `--force` option but just make it a bit smarter for certain cases? I mean, the fact that it was explicitly called is already an indication that I want meson to jfdi please, arguably ;)

**Comment by xclaesse** [comment]
When I wrote `meson subprojects update` I had my workflow in mind, selfishly:
- Clone gst-build master.
- Branch e.g. gst-plugins-good to work on a feature.
- `meson subprojects update` would rebase my feature branch on top of master, and if it has uncommitted changes it stash them before the rebase and re-apply.

That being said, that could not be the best workflow and we could change stuff, just keep in mind that subprojects could have local changes.

**Comment by tp-m** [comment]
>  just keep in mind that subprojects could have local changes.

Yeah, should stash those and let the user know if there are local changes.

**Comment by xclaesse** [comment]
wait, actually there is an option for this already: `meson subprojects update --rebase`, I think it does what you want, right?

**Comment by xclaesse** [comment]
Hm, no, it's not exactly that if current branch is 2.60.4 you don't want to rebase that on top of 2.64.0, but meson has no way to know if '2.60.4' is a personal feature branch that needs to be rebased, or an old upstream stable branch that should be abandoned and force reset on 2.64.0.

Maybe I'm a bit confused, but then I think --force is the only way.

**Comment by nirbheek** [comment]
I would prefer `--sync-revision`, since that's more descriptive.

**Comment by xclaesse** [comment]
I just faced a similar use-case today:
- In our CI we cache gst-build subprojects in docker images by doing `meson subprojects download` in Dockerfile. This avoids fetching them from scratch for each job.
- We have our own dev branch set in wrap files for some subprojects that we rebase on top of upstream branch times to times.
- The CI job script does `meson subprojects update` to fetch latest changes. But when we do a rebase, that fails because it's not fast forward, so we have to regenerate our docker images at that point, or work around it with custom `meson subproject foreach` invocation.

Basically what I need is something like `git fetch && git reset --hard <revision-from-wrap>`, because in that case we want to be sure we are at exactly that revision and we are sure to not have any useful local changes.

**Comment by xclaesse** [comment]
Wondering if it should be `meson subprojects checkout --update` maybe? Since the original use-case is about switching branch... dunno.

**Comment by nirbheek** [comment]
> Wondering if it should be `meson subprojects checkout --update` maybe? Since the original use-case is about switching branch... dunno.

It's about updating, so it should be part of the `meson subprojects update` workflow, I think.

**Comment by xclaesse** [comment]
or a new `meson subprojects reset`? Just throwing ideas to see what sticks, I don't really have a preference :p

**Comment by xclaesse** [comment]
I opted picked `meson subprojects update --reset` for now and reworked that command completely: https://github.com/mesonbuild/meson/pull/7564. Let me know what you think.
